experimental/segmenter/src/math_helper.rs

```diff
@@ -2,59 +2,338 @@
+}
+
+impl<'a> MatrixBorrowedMut<'a, 1> {
+    /// Calculate the dot product of a and b, adding the result to self.
+    ///
+    /// Note: For better dot product efficiency, if `b` is MxN, then `a` should be N;
+    /// this is the opposite of standard practice.
+    pub fn add_dot_2d(&mut self, a: MatrixBorrowed<1>, b: MatrixBorrowed<2>) {
+        let m = a.dim();
+        let n = self.as_borrowed().dim();
+        debug_assert_eq!(m, b.dim().1);
+        debug_assert_eq!(n, b.dim().0);
+        for i in 0..n {
+            if let (Some(dest), Some(b_sub)) = (self.as_mut_slice().get_mut(i), b.submatrix::<1>(i))
+            {
+                *dest += unrolled_dot(a.data, b_sub.data);
+            } else {
+                debug_assert!(false);
+            }
+        }
+    }
+}
+
+impl<'a> MatrixBorrowedMut<'a, 2> {
+    /// Calculate the dot product of a and b, adding the result to self.
+    ///
+    /// Self should be _MxN_; `a`, _O_; and `b`, _MxNxO_.
+    pub fn add_dot_3d(&mut self, a: MatrixBorrowed<1>, b: MatrixBorrowed<3>) {
+        let m = a.dim();
+        let n = self.as_borrowed().dim().0 * self.as_borrowed().dim().1;
+        debug_assert_eq!(m, b.dim().2);
+        debug_assert_eq!(n, b.dim().0 * b.dim().1);
+        // Note: The following two loops are equivalent, but the second has more opportunity for
+        // vectorization since it allows the vectorization to span submatrices.
+        // for i in 0..b.dim().0 {
+        //     self.submatrix_mut::<1>(i).add_dot_2d(a, b.submatrix(i));
+        // }
+        let lhs = a.as_slice();
+        for i in 0..n {
+            if let (Some(dest), Some(rhs)) = (
+                self.as_mut_slice().get_mut(i),
+                b.as_slice().get(i * m..(i + 1) * m),
+            ) {
+                *dest += unrolled_dot(lhs, rhs);
+            } else {
+                debug_assert!(false);
+            }
+        }
+    }
+}
+
+// Polyfill float operations with libm in case we're no_std.
+#[allow(unused_imports)]
+use num_traits::Float;
+
+/// `tanh` computes the tanh function for a scalar value.
+#[inline]
+pub fn tanh(x: f32) -> f32 {
+    x.tanh()
+}
+
+/// `sigmoid` computes the sigmoid function for a scalar value.
+#[inline]
+pub fn sigmoid(x: f32) -> f32 {
+    1.0 / (1.0 + (-x).exp())
+}
+
+/// Compute the dot product.
+///
+/// `xs` and `ys` must be the same length
+///
+/// (From ndarray 0.15.6)
+#[allow(clippy::indexing_slicing)] // all indexing is protected by the entry assertion
+fn unrolled_dot(xs: &[f32], ys: &[f32]) -> f32 {
+    debug_assert_eq!(xs.len(), ys.len());
+    if xs.len() != ys.len() {
+        return 0.0;
+    }
+    // eightfold unrolled so that floating point can be vectorized
+    // (even with strict floating point accuracy semantics)
+    let len = core::cmp::min(xs.len(), ys.len());
+    let mut xs = &xs[..len];
+    let mut ys = &ys[..len];
+    let mut sum = 0.0;
+    let (mut p0, mut p1, mut p2, mut p3, mut p4, mut p5, mut p6, mut p7) =
```

Comment:
suggestion: could probably be shorter with `let [mut p0, mut p1, ... mut p7] = [0; 8]`

Reply:
(this comment is against an out-of-date version of the file which I fixed according to similar feedback from Robert)

Reply:
I ran the end-to-end benchmark both ways

| Case | Run 1 | Run 2 | Run 3 |
|---|---|---|---|
| Tuple | 480.89 µs | 480.22 µs | 466.71 µs |
| Array | 485.55 µs | 491.95 µs | 487.85 µs |

It's not significant enough for me to say that one is faster than the other, but I will say that I prefer the tuple approach because (1) it avoids the risk of bounds checks and (2) it allows the compiler to pick the best memory layout.